### PR TITLE
Prep v0.7

### PR DESCRIPTION
**Pull request by williballenthin**
i've been working over here but might as well put it all in master

@mr-tz PR here as an FYI

adds:
  - tests
  - mypy
  - black, sort

fixes:
  - various APIs that didn't work as we thought

changes:
  - py3 only
  - remove emu arg from monitor ctor
  - make hooks simply a callable, like a function or class with `__call__`. no longer a subclass of `Hook`
  - snake case method names
  - make dbg driver breakpoint API via property access, not `addBreakpoint`
  - dbg driver `run_to_*` raises exceptions more reliably, probably need to handle them in client code

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -66,3 +66,4 @@ target/
 .ipynb_checkpoints
 >>>>>>> 5e221b7233993e6b9c661f6c6b37c733cddd4ea9
```
Comment (williballenthin) [inline .gitignore:67]: TODO: fix merge here

File: tests/test_cfg.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    assert int(cfg.get_root_basic_block()) == 0x10001010
+
+    tails = list(cfg.get_leaf_basic_blocks())
+    assert len(tails) == 1
+
+    tail = tails[0]
+    assert int(tail) == 0x100011E8
+
+    assert len(list(cfg.get_predecessor_basic_blocks(tail))) == 4
```
Comment (williballenthin) [inline tests/test_cfg.py:48]: this function actually didn't work 🤦🏼

File: viv_utils/__init__.py [hunk 3]
```diff
@@ -114,60 +124,31 @@ def getWorkspace(fp, analyze=True, reanalyze=False, verbose=False, should_save=T
     return vw
 
 
-class LoggingObject(object):
```
Comment (williballenthin) [inline viv_utils/__init__.py:117]: i think this object was ultimately a bad idea. well, it did work, but i question its performance, and more importantly, it discourages from using a single global logger like all the rest of our code.

File: viv_utils/emulator_drivers.py [hunk 4]
```diff
@@ -31,67 +30,39 @@ def __str__(self):
...
-
+class Hook:
     def hook(self, callname, emu, callconv, api, argv):
         # must return something other than None if handled
         # raise UnsupportedFunction to pass
         raise UnsupportedFunction()
 
 
-class Monitor(vivisect.impemu.monitor.EmulationMonitor, LoggingObject):
-    def __init__(self, vw):
```
Comment (williballenthin) [inline viv_utils/emulator_drivers.py:45]: breaking change: monitor no longer takes the `vw` (access via `emu.vw`)

File: viv_utils/emulator_drivers.py [hunk 5]
```diff
@@ -31,67 +30,39 @@ def __str__(self):
...
-
-    def getStackValue(self, emu, offset):
-        return emu.readMemoryFormat(emu.getStackCounter() + offset, "<P")[0]
-
-    def dumpStack(self, emu, num):
-        self._logger.debug("stack: ESP: %s", hex(emu.getStackCounter()))
-        for i in range(num):
-            self.d("stack: ESP + %s: %s",
-                               hex(emu.imem_psize * i),
-                               hex(self.getStackValue(emu, emu.imem_psize * i)))
```
Comment (williballenthin) [inline viv_utils/emulator_drivers.py:59]: old code: nobody used this

File: viv_utils/__init__.py (hunk 3 above)
Comment (mr-tz) [reply viv_utils/__init__.py:117]: it was useful at one point, but yeah agreed!

File: viv_utils/emulator_drivers.py (hunk 4 above)
Comment (mr-tz) [reply viv_utils/emulator_drivers.py:45]: gotta take note of that!

File: viv_utils/emulator_drivers.py (hunk 5 above)
Comment (mr-tz) [reply viv_utils/emulator_drivers.py:59]: yep

Comment (mr-tz) [reply viv_utils/emulator_drivers.py:59]: wait, I think we do in FLOSS

Comment (mr-tz) [reply viv_utils/emulator_drivers.py:59]: yes, but can remove it there or move it, if needed

File: tests/test_cfg.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    # four xrefs: fallthrough and three jumps
+    assert len(list(viv_utils.get_all_xrefs_to(pma01, 0x100011E8))) == 4
+
+
+def test_cfg(pma01):
+    f = viv_utils.Function(pma01, 0x10001010)
+    cfg = viv_utils.CFG(f)
+
+    assert int(cfg.get_root_basic_block()) == 0x10001010
```
Comment (mr-tz) [inline tests/test_cfg.py:40]: note: `get_root_basic_block` can raise `KeyError` for some functions

File: viv_utils/emulator_drivers.py [hunk 7]
```diff
@@ -182,43 +155,41 @@ def doHook(self, pc, op):
...
                 ret = hook.hook(callname, self, callconv, api, argv)
                 # take the first result
                 # not ideal, but works in the common case
                 if ret is not None:
-                    self._logger.debug("driver hook handled call: %s", callname)
+                    logger.debug("driver: hook handled call: %s", callname)
                     return True
+
+                if callname and callname not in ("UnknownApi", "?"):
+                    logger.debug("driver: hook API call NOT handled: %s", callname)
```
Comment (mr-tz) [inline viv_utils/emulator_drivers.py:175]: when done here this triggers on every hook iteration, even though a later hook may have actually handled the call

File: viv_utils/emulator_drivers.py [hunk 8]
```diff
@@ -418,7 +393,7 @@ def _runFunction(self, funcva, stopva=None, maxhit=None, maxloop=None, maxrep=No
 
             # Check if we are beyond our loop max...
             if maxloop != None:
-                lcount = vg_path.getPathLoopCount(self.curpath, 'bva', va)
+                lcount = vg_path.getPathLoopCount(self.curpath, "bva", va)
```
Comment (mr-tz) [inline viv_utils/emulator_drivers.py:396]: not sure this works or I don't know how to use it :)

**Review by mr-tz — COMMENTED** [review]
nice, looking forward to the updates!

File: viv_utils/emulator_drivers.py [hunk 9]
```diff
@@ -30,11 +28,50 @@ def __str__(self):
...
+# a hook is a callable, such as a function or class with `__call__`,
+# if the hook is "stateless", use a simple function:
+#
+#     hook_OutputDebugString(name, emu, cconv, api, argv):
+#         if name != "kernel32.OutputDebugString": return False
+#         logger.debug("OutputDebugString: %s", emu.readString(argv[0]))
+#         cconv.execCallReturn(emu, 0, len(argv))
+#         return True
+#
+# if the hook is "stateful", such as a hook that records arguments, use a class:
```
Comment (mr-tz) [inline viv_utils/emulator_drivers.py:62]: neat!

File: viv_utils/emulator_drivers.py [hunk 10]
```diff
@@ -152,46 +219,41 @@ def doHook(self, pc, op):
         for mon in self._monitors:
             try:
                 r = mon.apicall(self, op, pc, api, argv)
-                if r is not None:
-                    # take the first result
-                    # not ideal, but works in the common case
-                    logger.debug("driver: monitor hook handled call: %s", callname)
-                    return True
-            except StopEmulation:
-                raise
```
Comment (mr-tz) [inline viv_utils/emulator_drivers.py:188]: will this still be available to higher-level callers?

File: viv_utils/emulator_drivers.py [hunk 11]
```diff
@@ -31,75 +28,136 @@ def __str__(self):
...
+#         def __init__(self):
+#             self.paths = set()
+#
+#         def __call__(self, name, emu, cconv, api, argv):
+#             if name != "kernel32.CreateFileA": return False
+#             self.paths.add(emu.readString(argv[0]))
+#             cconv.execCallReturn(emu, 0, len(argv))
+#             return True
+#
+Hook = Callable[[FunctionName, Emulator, CallingConvention, API, List[int]], bool]
```
Comment (williballenthin) [inline viv_utils/emulator_drivers.py:74]: breaking change

File: viv_utils/emulator_drivers.py (hunk 10 above)
Comment (williballenthin) [reply viv_utils/emulator_drivers.py:188]: i messed up a merge and forgot this, sorry

Comment (williballenthin) [reply viv_utils/emulator_drivers.py]: re-added in 5df84599ee18f4ec7deddf43f3b8926d01a951b8

File: viv_utils/emulator_drivers.py [hunk 12]
```diff
@@ -31,141 +28,178 @@ def __str__(self):
...
+#             self.paths = set()
+#
+#         def __call__(self, emu, api, argv):
+#             _, _, cconv, name, _ = api
+#             if name != "kernel32.CreateFileA": return False
+#             self.paths.add(emu.readString(argv[0]))
+#             cconv.execCallReturn(emu, 0, len(argv))
+#             return True
+#
+Hook = Callable[[Emulator, API, List[int]], bool]
```
Comment (mr-tz) [inline viv_utils/emulator_drivers.py:77]: How about adding a `NopHook` helper or something? FLOSS could use that quite a bit (see e.g. https://github.com/mandiant/flare-floss/pull/499/files#diff-75983f483110f7478f670c58faccc16218afe8aae2469fe8a8c2d522e1d00e54R397-R403) but we could also add it there.

File: viv_utils/emulator_drivers.py (hunk 7 above)
Comment (williballenthin) [reply viv_utils/emulator_drivers.py]: fixed

File: tests/test_cfg.py (hunk 6 above)
Comment (williballenthin) [reply tests/test_cfg.py:40]: can you share the hash of a sample that shows this behavior? i'd like to try to fix it.

File: viv_utils/emulator_drivers.py [hunk 13]
```diff
@@ -31,141 +32,188 @@ def __str__(self):
...
     def posthook(self, emu, op, endpc):
-        #self.dumpStack(emu, 4)
-        #self._logger.debug("  EBX: %s", hex(emu.getRegisterByName("ebx")))
         pass
 
-    def apicall(self, driver, op, pc, api, argv):
-        #self._logger.debug("apicall: %s %s %s %s", op, pc, api, argv)
-        # if non-None is returned, then it signals that the API call was handled
-        #   and this function *must* handle cleaning up the stack
+    def preblock(self, emu, blockstart):
```
Comment (mr-tz) [inline viv_utils/emulator_drivers.py:93]: this now differs from vivisect def, may be an issue?
```
def apicall(self, emu, op, pc, api, argv):
```

File: tests/test_driver.py [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+
+class CreateMutexAHook:
+    """capture the mutex names passed to CreateMutexA"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.mutexes = set()
+
+    def __call__(self, emu, api, argv):
+        _, _, cconv, name, _ = api
```
Comment (mr-tz) [inline tests/test_driver.py:92]: (1) would be helpful to have some helper/magic for this common pattern and the one below, see (2)

File: tests/test_driver.py [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+
+        if name != "kernel32.CreateMutexA":
+            return
+
+        mutex = emu.readString(argv[2])
+        self.mutexes.add(mutex)
+
+        cconv = emu.getCallingConvention(cconv)
+        cconv.execCallReturn(emu, 0, len(argv))
```
Comment (mr-tz) [inline tests/test_driver.py:101]: (2) see (1) 😄

File: viv_utils/emulator_drivers.py [hunk 16]
```diff
@@ -178,131 +226,226 @@ def doHook(self, pc, op):
...
         if callname in emu.hooks:
+            # this is where vivisect-internal hooks are stored,
+            # such as those provided by impapi.
+            # note that we prefer locally configured hooks, first.
             hook = emu.hooks.get(callname)
             try:
-                hook(self, callconv, api, argv)
-                self._logger.debug("emu hook handled call: %s", callname)
-                return True
+                hook(self, api, argv)
```
Comment (mr-tz) [inline viv_utils/emulator_drivers.py:261]: vivisect defines hook with a different API, e.g. `def kernel32_GetModuleHandleA(self, emu, callconv, api, argv)`

should we keep the hook API consistent or pass the modified arguments here?

**Pull request merged**